partitura/performance.py

```diff
@@ -280,6 +284,127 @@ def adjust_offsets_w_sustain(
+
+    def __str__(self):
+        return f"PerformedNote: {self['id']}"
+
+    def __eq__(self, other):
+        if not isinstance(PerformedNote):
+            return False
+        if not self.keys() == other.keys():
+            return False
+        return np.all(
+            np.array([self[k] == other[k] for k in self.keys() if k in other.keys()])
+        )
+
+    def __hash__(self):
+        return hash(self["id"])
+
+    def __lt__(self, other):
+        return self["note_on"] < other["note_on"]
+
+    def __le__(self, other):
+        return self["note_on"] <= other["note_on"]
+
+    def __gt__(self, other):
+        return self["note_on"] > other["note_on"]
+
+    def __ge__(self, other):
+        return self["note_on"] >= other["note_on"]
+
+    def __getitem__(self, key):
+        return self.get(key, None)
+
+    def _setitem_new(self, key, value):
+        if key not in self._accepted_keys:
+            raise KeyError(f"Key {key} not in PerformedNote")
+        elif key == "note_off":
+            # Verify that the note_off is after the note_on
+            if value < self["note_on"]:
+                raise ValueError(f"note_off must be after or equal to note_on")
+            self["sound_off"] = (
+                value if self["sound_off"] < value else self["sound_off"]
+            )
+            self["note_off"] = value
+        elif key == "note_on":
+            # Verify that the note_on is before the note_off
+            if value > self["note_off"]:
+                raise ValueError(f"note_on must be before or equal to note_off")
+
+            self["duration_sec"] = self["note_off"] - value
+            self["note_on"] = value
+        elif key == "sound_off":
+            # Verify that the sound_off is after the note_off
+            if value <= self["note_off"]:
+                raise ValueError(f"sound_off must be after or equal to note_off")
+            self["sound_off"] = value
+        else:
+            self[key] = value
+
+    def __delitem__(self, key):
+        raise KeyError("Cannot delete items from PerformedNote")
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def __len__(self):
+        return len(self.keys())
+
+    def __contains__(self, key):
+        return key in self.keys()
+
+    def _validate_values(self):
+        if self["pitch"] > 127 or self["pitch"] < 0:
+            raise ValueError(f"pitch must be between 0 and 127")
+        if self["note_on"] < 0:
+            raise ValueError(
+                f"Note on value provided is invalid, must be greater than or equal to 0"
+            )
+        if self["note_off"] < 0 or self["note_off"] < self["note_on"]:
+            raise ValueError(
+                f"Note off value provided is invalid, "
+                f"must be greater than or equal to 0 and greater or equal to note_on"
+            )
+        if self["velocity"] > 127 or self["velocity"] < 0:
+            raise ValueError(f"velocity must be between 0 and 127")
+
```

Comment:
is this the solution you want? the same checks are made in the `__setitem__` , could they not be reused? the whole thing seems a bit hacky to me